src/analysisd/alerts/exec.c

```diff
@@ -86,68 +63,93 @@ void OS_Exec(int execq, int *arq, const Eventinfo *lf, const active_response *ar
         if (!(Config.ar & LOCAL_AR)) {
             goto cleanup;
         }
+        if(getActiveResponseInJSON(lf, ar, exec_msg))
+            goto cleanup;
 
-        snprintf(exec_msg, OS_SIZE_1024,
-                 "%s %s %s %ld.%ld %d %s %s %s",
-                 ar->name,
-                 user,
-                 ip,
-                 (long int)lf->time.tv_sec,
-                 __crt_ftell,
-                 lf->generated_rule->sigid,
-                 lf->location,
-                 filename ? filename : "-",
-                 extra_args ? extra_args : "-");
-
-        if (OS_SendUnix(execq, exec_msg, 0) < 0) {
+        if (OS_SendUnix(execq, exec_msg, 0) < 0)
             merror("Error communicating with execd.");
-        }
     }
 
     /* Active Response to the forwarder */
     else if ((Config.ar & REMOTE_AR)) {
-        int rc;
-        /* If lf->location start with a ( was generated by remote agent and its
-         * ID is included in lf->location if missing then it must have been
-         * generated by the local analysisd, so prepend a false id tag */
-        if (lf->location[0] == '(') {
-            snprintf(exec_msg, OS_SIZE_1024,
-                     "%s %c%c%c %s %s %s %s %ld.%ld %d %s %s %s",
-                     lf->location,
-                     (ar->location & ALL_AGENTS) ? ALL_AGENTS_C : NONE_C,
-                     (ar->location & REMOTE_AGENT) ? REMOTE_AGENT_C : NONE_C,
-                     (ar->location & SPECIFIC_AGENT) ? SPECIFIC_AGENT_C : NONE_C,
-                     ar->agent_id != NULL ? ar->agent_id : "(null)",
-                     ar->name,
-                     user,
-                     ip,
-                     (long int)lf->time.tv_sec,
-                     __crt_ftell,
-                     lf->generated_rule->sigid,
-                     lf->location,
-                     filename ? filename : "-",
-                     extra_args ? extra_args : "-");
+
+        /* We need to identify the version of Wazuh installed on the agents because 
+         * if the agents version is <4.2, it is necessary to send legacy 
+         * message in string format. Instead if the agent version is >=4.2, 
+         * we need to send the new message in JSON format.*/
+
+        if (ar->location & ALL_AGENTS) {
```

Comment:
I see a problem in this if. You are not changing the `ar->location` from `ALL_AGENTS` to `SPECIFIC_AGENT` for every single agent.

Reply:
ar->location is a pointer to constant, my idea is to modify the function get_exec_msg in the following way to solve the problem:

```
snprintf(temp_msg, OS_SIZE_1024,
            "%s %c%c%c %s",
            lf->location,
            NONE_C,
            (ar->location & REMOTE_AGENT) ? REMOTE_AGENT_C : NONE_C,
            (ar->location & SPECIFIC_AGENT || ar->location & ALL_AGENTS) ? SPECIFIC_AGENT_C : NONE_C,
            ar->agent_id != NULL ? ar->agent_id : "(null)");

    strcat(exec_msg, temp_msg);
```